src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -266,147 +232,94 @@ public override Task CloseAsync(WebSocketCloseStatus closeStatus, string? status
-            {
-                lock (self._thisLock)
+                if (_disposed)
                 {
-                    AbortCore(self, self.State);
+                    return;
                 }
-            }, this);
-#else
-            AbortCore(this, this.State);
-#endif
+                state = State;
+            } // lock
+            AbortCore(this, state);
         }
 
         public override void Dispose()
         {
-#if FEATURE_WASM_THREADS
-            if (_disposed)
-            {
-                return;
-            }
-            _innerWebSocket?.SynchronizationContext.Send(static (BrowserWebSocket self) =>
+            WebSocketState state;
+            lock (_lockObject)
             {
-                lock (self._thisLock)
+                if (_disposed)
                 {
-                    DisposeCore(self);
+                    return;
                 }
-            }, this);
-#else
-            DisposeCore(this);
-#endif
-        }
+                _disposed = true;
+                state = State;
+            } // lock
 
-        private void ThrowIfDisposed()
-        {
-#if FEATURE_WASM_THREADS
-            lock (_thisLock)
+            if (state < WebSocketState.Closed && state != WebSocketState.None)
             {
-#endif
-                ObjectDisposedException.ThrowIf(_disposed, this);
-#if FEATURE_WASM_THREADS
-            } //lock
-#endif
+                AbortCore(this, state);
+            }
+            if (state != WebSocketState.Aborted)
+            {
+                FastState = WebSocketState.Closed;
+            }
+            _innerWebSocket?.Dispose();
+            responseStatusHandle?.Dispose();
         }
 
-        #region methods always called on one thread only, exclusively
-
-        private static void DisposeCore(BrowserWebSocket self)
+        private void ThrowIfDisposed()
         {
-            if (!self._disposed)
+            lock (_lockObject)
             {
-                var state = self.State;
-                self._disposed = true;
-                if (state < WebSocketState.Aborted && state != WebSocketState.None)
-                {
-                    AbortCore(self, state);
-                }
-                if (self.FastState != WebSocketState.Aborted)
-                {
-                    self.FastState = WebSocketState.Closed;
-                }
-                self._innerWebSocket?.Dispose();
-                self._innerWebSocket = null;
-                self.responseStatusHandle?.Dispose();
-            }
+                ObjectDisposedException.ThrowIf(_disposed, this);
+            } // lock
         }
 
         private static void AbortCore(BrowserWebSocket self, WebSocketState currentState)
         {
-            if (!self._disposed && currentState != WebSocketState.Closed)
+            lock (self._lockObject)
             {
                 self.FastState = WebSocketState.Aborted;
```

Comment:
i think it's incorrect that we are able to change the state or aborted flag if we're already disposed. should it be possible to change internal state on a disposed socket?

Reply:
I updated the code. We have now 2 abort scenarios
- called directly, which will do blocking abort
- called from `Dispose`, which will schedule the abort onto the owner thread. In this case `_aborted` will change after `_disposed`